### PR TITLE
Sneddon manufactured solution.

### PR DESCRIPTION
**Pull request by isakhammer**
## Proposed changes

This pull request improves upon the previously submitted PR (https://github.com/pmgbergen/porepy/pull/1283). Runtime on my computer is approx 12 seconds.

This pull request includes reimplementing Sneddon's analytical solution as outlined in https://doi.org/10.1007/s10596-020-10002-5. Sneddon's solution provides an analytical method to compute the normal jump displacement for a pressurized fracture under linear elasticity conditions. The boundary element method is then integrated to compute semi-analytical boundary values, which are subsequently used to verify the MPSA numerical method for an open fracture. This process validates the accuracy and convergence of the proposed numerical method, making it well-suited as a test case.

## Types of changes

What types of changes does this PR introduce to PorePy?
_Put an `x` in the boxes that apply._

- [ ] Minor change (e.g., dependency bumps, broken links).
- [ ] Bugfix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [x] Testing (contribution related to testing of existing or new functionality).
- [ ] Documentation (contribution related to adding, improving, or fixing documentation).
- [ ] Maintenance (e.g., improve logic and performance, remove obsolete code).
- [ ] Other:

## Checklist

_Put an `x` in the boxes that apply or explain briefly why the box is not relevant._

- [x] The documentation is up-to-date.
- [x] Static typing is included in the update.
- [x] This PR does not duplicate existing functionality.
- [x] The update is covered by the test suite (including tests added in the PR).
- [ ] If new skipped tests have been introduced in this PR, `pytest` was run with the `--run-skipped` flag.

### CODE REVIEW

File: tests/functional/setups/manu_sneddon_2d.py [hunk 1]
```diff
@@ -0,0 +1,436 @@
...
+        h: bem segment length
+        n: number of bem segments
+        theta: orientation of the fracture
+        center: center of the fracture
+
+    Return:
+        Array of BEM centers
+    """
+
+    # Coordinate system (4.5.1) in page 57 in in book Crouch Starfield 1983 Boundary Element Methods in Solid Mechanics
```
Comment (IvarStefansson) [inline tests/functional/setups/manu_sneddon_2d.py:43]: Move reference (top of file?) and cite with a shorthand

File: tests/functional/setups/manu_sneddon_2d.py [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+
+    Parameter:
+        a: half fracture length
+        eta: distance from fracture centre
+        p0: pressure
+        G: shear modulus
+        poi: poisson ratio
+
+    Return
+        List of analytical normal displacement jumps.
```
Comment (IvarStefansson) [inline tests/functional/setups/manu_sneddon_2d.py:71]: Not a list

File: tests/functional/setups/manu_sneddon_2d.py [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+        Semi-analytical boundary displacement values
+    """
+
+    bc_val = np.zeros((g.dim, g.num_faces))
+
+    alpha = np.pi / 2 - theta
+
+    bound_face_centers = g.face_centers[:, bound_faces]
+
+    for i in range(0, u_a.size):
```
Comment (IvarStefansson) [inline tests/functional/setups/manu_sneddon_2d.py:198]: Add some documentation.

File: tests/functional/setups/manu_sneddon_2d.py [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+class ManuExactSneddon2dSetup:
+    """
+    Class for setting up the analytical solution for the pressurized fracture problem.
+    
+    
+    """
+  
+    def __init__(self, setup):
+        # Initialize private variables from the setup dictionary
+        self.p0 = setup.get("p0")
```
Comment (IvarStefansson) [inline tests/functional/setups/manu_sneddon_2d.py:222]: Why not just use setup.p0 etc.? We could also access as self.setup.p0 in the rest of this class.

File: tests/functional/setups/manu_sneddon_2d.py [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+        eta = compute_eta(bem_centers, center)
+        
+        
+        u_a = -analytical_displacements(self.a, eta, self.p0, self.shear_modulus, self.poi)
+        u_bc = assign_bem(sd, h / 2, box_faces, self.theta, bem_centers, u_a, self.poi)
+        return u_bc
+    
+
+    def exact_sol_fracture( self,
+    gb: pp.MixedDimensionalGrid) -> tuple:
```
Comment (IvarStefansson) [inline tests/functional/setups/manu_sneddon_2d.py:257]: Change to mdg

File: tests/functional/setups/manu_sneddon_2d.py [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+        eta = compute_eta(fracture_faces, fracture_center)
+        apertures = analytical_displacements(self.a, eta, self.p0, self.shear_modulus, self.poi)
+
+        return eta, apertures
+
+
+
+
+
+class ModifiedGeometry(PorePyModel):
```
Comment (IvarStefansson) [inline tests/functional/setups/manu_sneddon_2d.py:292]: More precise class name.

File: tests/functional/setups/manu_sneddon_2d.py [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+    def set_domain(self) -> None:
+        """Defining a two-dimensional square domain with sidelengths length and height."""
```
Comment (IvarStefansson) [inline tests/functional/setups/manu_sneddon_2d.py:294]: If it's square, we only need one parameter. We can reuse a class from applications

File: tests/functional/setups/manu_sneddon_2d.py [hunk 8]
```diff
@@ -0,0 +1,436 @@
...
+            sd: Subdomain for which the boundary conditions are to be set.
+        Returns:
+            Boundary condition type for the problem.
+        """
+        bounds = self.domain_boundary_sides(sd)
+
+        # Set the type of west and east boundaries to Dirichlet. North and south are
+        # Neumann by default.
+        bc = pp.BoundaryConditionVectorial(sd, bounds.all_bf, "dir")
+        frac_face = sd.tags["fracture_faces"]
```
Comment (IvarStefansson) [inline tests/functional/setups/manu_sneddon_2d.py:329]: I think there's a function for this. internal_bc_to_dirichlet or whatever.

Comment (IvarStefansson) [reply tests/functional/setups/manu_sneddon_2d.py:329]: I also believe this to be the default behaviour - all Dirichlet on external and internal boundaries.

File: tests/functional/setups/manu_sneddon_2d.py [hunk 9]
```diff
@@ -0,0 +1,436 @@
...
+
+
+@dataclass
+class SneddonData:
+    """Data class for storing the error in the displacement field."""
+    error_displacement: pp.number
+
+class SneddonDataSaving(VerificationDataSaving):
+    """Class for saving the error in the displacement field."""
+    def collect_data(self):
```
Comment (IvarStefansson) [inline tests/functional/setups/manu_sneddon_2d.py:401]: typing

File: tests/functional/test_sneddon_2d.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+
+
+def test_order_of_convergence(
+    actual_ooc,
+) -> None:
+    """Test observed order of convergence.
+    """
+    # We  the order of L2 convergence on the fracture of displacement to be about 1.0 
+    assert 0.85 <   actual_ooc["ooc_displacement"]
```
Comment (IvarStefansson) [inline tests/functional/test_sneddon_2d.py:107]: The test needs to be more precise.

File: tests/functional/setups/manu_sneddon_2d.py [hunk 11]
```diff
@@ -1,128 +1,134 @@
...
+    # Terms in (7.4.6) in
+    # Crouch Starfield 1983 Boundary Element Methods in Solid Mechanics page 168
     x_bar[0, :] = (x[0, :] - xc[0]) * np.cos(alpha) + (x[1, :] - xc[1]) * np.sin(alpha)
     x_bar[1, :] = -(x[0, :] - xc[0]) * np.sin(alpha) + (x[1, :] - xc[1]) * np.cos(alpha)
     return x_bar
 
 
 def get_bc_val(
-    g: pp.Grid,
+    grid: pp.Grid,
```
Comment (IvarStefansson) [inline tests/functional/setups/manu_sneddon_2d.py:99]: sd?

File: tests/functional/setups/manu_sneddon_2d.py [hunk 12]
```diff
@@ -165,77 +171,71 @@ def get_bc_val(
...
-        self.poi = setup.get("poi")
-        self.length = setup.get("length")
-        self.height = setup.get("height")
-         
-  
-    def exact_sol_global(self, sd):
-        """
-        Compute the analytical solution for the pressurized fracture problem in question.
-        
+class ManuSneddonExactSolution2d:
```
Comment (IvarStefansson) [inline tests/functional/setups/manu_sneddon_2d.py:217]: Do we really want "Manu", since its an analytical, rather than manufactured, solution?

File: tests/functional/setups/manu_sneddon_2d.py [hunk 13]
```diff
@@ -328,106 +334,112 @@ def bc_type_mechanics(self, sd: pp.Grid) -> pp.BoundaryConditionVectorial:
...
-    pass
-    
+    """Complete Sneddon setup, including data collection and the analytical solution."""
+
+    exact_sol: ManuSneddonExactSolution2d
 
+    def set_materials(self):
+        """Initiating additionally the exact solution."""
+        super().set_materials()
+        self.exact_sol = ManuSneddonExactSolution2d(self.params)
```
Comment (IvarStefansson) [inline tests/functional/setups/manu_sneddon_2d.py:445]: Wrong place to set exact solution

File: tests/functional/setups/manu_sneddon_2d.py [hunk 14]
```diff
@@ -0,0 +1,445 @@
...
+
+    Return
+        Array containing the analytical normal displacement jumps.
+
+    """
+    cons = (1 - poi) / G * p0 * a * 2
+    return cons * np.sqrt(1 - np.power(eta / a, 2))
+
+
+def transform(xc: np.ndarray, x: np.ndarray, alpha: float) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (Yuriyzabegaev) [reply tests/functional/setups/manu_sneddon_2d.py:88]: I would rename this function, e.g. `tranform_boundary_face_coordinates_bem`

File: tests/functional/setups/manu_sneddon_2d.py [hunk 15]
```diff
@@ -0,0 +1,445 @@
...
+        bc_val += u_bound
+
+    return bc_val
+
+
+class ManuSneddonExactSolution2d:
+    """Class representing the analytical solution for the pressurized fracture problem."""
+
+    def __init__(self, params: dict):
+        self.p0 = params.get("p0")
```
Comment (Yuriyzabegaev) [inline tests/functional/setups/manu_sneddon_2d.py]: `dict.get` method returns `None` if there is no value for this key. This may lead to intricate errors to debug later. I suggest rewriting this with square brackets and add type hints.

File: tests/functional/setups/manu_sneddon_2d.py [hunk 16]
```diff
@@ -0,0 +1,445 @@
...
+    def exact_sol_global(self, sd: pp.Grid) -> np.ndarray:
+        """Compute the analytical solution for the pressurized fracture problem in
+        question.
+
+        Parameters:
+            sd: Subdomain for which the analytical solution is to be computed.
+
+        """
+
+        n = 1000
```
Comment (Yuriyzabegaev) [inline tests/functional/setups/manu_sneddon_2d.py:242]: This `n` is a number of bem segments. Should it be hardcoded here? Will it not increase with grid refinement?

File: tests/functional/setups/manu_sneddon_2d.py [hunk 17]
```diff
@@ -0,0 +1,445 @@
...
+        )
+
+    def grid_type(self) -> Literal["simplex", "cartesian", "tensor_grid"]:
+        """Choosing the grid type for our domain."""
+        return self.params.get("grid_type", "simplex")
+
+    def set_fractures(self):
+        """Setting a single line fracture in the domain using ``params['frac_pts']``."""
+
+        points = self.params["frac_pts"]
```
Comment (Yuriyzabegaev) [inline tests/functional/setups/manu_sneddon_2d.py]: If the user provides a fracture which does not cross the domain center, the assumptions made in the analytical solution computation get invalidated. Should this be removed from the parameters and instead be hardcoded here?

File: tests/functional/test_sneddon_2d.py [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+    )
+
+    # Calculate and return the order of convergence for the displacement
+    order_dict = conv_analysis.order_of_convergence(
+        conv_analysis.run_analysis(), data_range=slice(None, None, None)
+    )
+    return order_dict
+
+
+def test_order_of_convergence(actual_ooc: dict) -> None:
```
Comment (Yuriyzabegaev) [inline tests/functional/test_sneddon_2d.py:104]: What is the reason why the body of the test is moved to a fixture? I don't mind, but curious.

File: tests/functional/test_sneddon_2d.py [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+poi = 0.25
+shear_modulus = 1
+lam = (
+    2 * shear_modulus * poi / (1 - 2 * poi)
+)  # Convertion formula from shear modulus and poission to lame lambda parameter
+
+solid = pp.SolidConstants(shear_modulus=shear_modulus, lame_lambda=lam)
+
+
+def compute_frac_pts(
```
Comment (Yuriyzabegaev) [inline tests/functional/test_sneddon_2d.py:21]: Back to my comment about the storing `"frac_pts"` in `params`. This function feels like a perfect complement to the Sneddon model setup, because it generates a fracture that satisfies the Sneddon model assumption. The user can pass the parameters to edit the fracture length and rotation etc. I suggest moving it there.

File: tests/functional/setups/manu_sneddon_2d.py (hunk 1 above)
Comment (vlipovac) [reply tests/functional/setups/manu_sneddon_2d.py]: All references moved to module doc string, referencing in inline comments reworked.

File: tests/functional/setups/manu_sneddon_2d.py (hunk 7 above)
Comment (vlipovac) [reply tests/functional/setups/manu_sneddon_2d.py]: Switched to 'SquareDomainOrthogonalFractures', but I had to overwrite the fractures to fit the setup.

File: tests/functional/setups/manu_sneddon_2d.py (hunk 13 above)
Comment (vlipovac) [reply tests/functional/setups/manu_sneddon_2d.py]: Moved to __init__

File: tests/functional/setups/manu_sneddon_2d.py [hunk 20]
```diff
@@ -0,0 +1,425 @@
...
+
+import porepy as pp
+from porepy.applications.convergence_analysis import ConvergenceAnalysis
+from porepy.applications.md_grids.model_geometries import (
+    SquareDomainOrthogonalFractures,
+)
+from porepy.geometry.distances import point_pointset
+
+
+def compute_eta(pointset_centers: np.ndarray, center: np.ndarray) -> np.ndarray:
```
Comment (vlipovac) [inline tests/functional/setups/manu_sneddon_2d.py:25]: This is a shallow wrapper which can be eliminated.

File: tests/functional/setups/manu_sneddon_2d.py [hunk 21]
```diff
@@ -0,0 +1,425 @@
...
+    x_0 = center[0] - (a - 0.5 * h) * np.sin(alpha)
+    y_0 = center[1] - (a - 0.5 * h) * np.cos(alpha)
+    for i in range(0, n):
+        bem_centers[0, i] = x_0 + i * h * np.sin(alpha)
+        bem_centers[1, i] = y_0 + i * h * np.cos(alpha)
+
+    return bem_centers
+
+
+def analytical_displacements(
```
Comment (vlipovac) [inline tests/functional/setups/manu_sneddon_2d.py:67]: Should be moved to class representing the analytical solution.

File: tests/functional/setups/manu_sneddon_2d.py [hunk 22]
```diff
@@ -0,0 +1,425 @@
...
+    Return
+        Array containing the analytical normal displacement jumps.
+
+    """
+    # See [2], p. 425 equ. 92
+    cons = (1 - poi) / G * p0 * a * 2
+    return cons * np.sqrt(1 - np.power(eta / a, 2))
+
+
+def transform(xc: np.ndarray, x: np.ndarray, alpha: float) -> np.ndarray:
```
Comment (vlipovac) [inline tests/functional/setups/manu_sneddon_2d.py:88]: Only used once in below method, can be eliminated and code put there.
In general, all these helper methods before the classes are created seem to be part of the analytical solution and can be moved there. Would also help reduce the signatures.

File: tests/functional/setups/manu_sneddon_2d.py [hunk 23]
```diff
@@ -0,0 +1,445 @@
...
+
+    """
+
+    # Coordinate system (4.5.1) in page 57 in in book
+    # Crouch Starfield 1983 Boundary Element Methods in Solid Mechanics
+
+    bem_centers = np.zeros((3, n))
+    x_0 = center[0] - (a - 0.5 * h) * np.sin(alpha)
+    y_0 = center[1] - (a - 0.5 * h) * np.cos(alpha)
+    for i in range(0, n):
```
[the remark this answers is not in the document]
Comment (IvarStefansson) [reply tests/functional/setups/manu_sneddon_2d.py:60]: Yes, please!

File: tests/functional/setups/manu_sneddon_2d.py (hunk 21 above)
Comment (IvarStefansson) [reply tests/functional/setups/manu_sneddon_2d.py:67]: I agree

File: tests/functional/setups/manu_sneddon_2d.py (hunk 22 above)
Comment (IvarStefansson) [reply tests/functional/setups/manu_sneddon_2d.py:88]: Good observation. Please implement!

File: tests/functional/setups/manu_sneddon_2d.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (IvarStefansson) [reply tests/functional/setups/manu_sneddon_2d.py:88]: The name may be improved, but the exact name may depend on context, i.e., Veljko's suggestion to move it.

File: tests/functional/setups/manu_sneddon_2d.py (hunk 16 above)
Comment (IvarStefansson) [reply tests/functional/setups/manu_sneddon_2d.py:242]: You're right, it should not be hardcoded here. Perhaps retrieve from params as e.g. "num_bem_segments"?

File: tests/functional/setups/manu_sneddon_2d.py [hunk 24]
```diff
@@ -0,0 +1,425 @@
...
+    displacement_jump: Callable[[list[pp.Grid]], pp.ad.Operator]
+
+    def collect_data(self) -> ManuSneddonSaveData:
+        """Collecting the error in the displacement field."""
+        frac_sd = self.mdg.subdomains(dim=self.nd - 1)
+        nd_vec_to_normal = self.normal_component(frac_sd)
+
+        # Computing the numerical displacement jump along the fracture on the fracture
+        # cell centers.
+        u_n: pp.ad.Operator = nd_vec_to_normal @ self.displacement_jump(frac_sd)
```
[the remark this answers is not in the document]
Comment (IvarStefansson) [reply tests/functional/setups/manu_sneddon_2d.py:365]: Oh, dear. Rename below to u_n_value?

File: tests/functional/setups/manu_sneddon_2d.py [hunk 25]
```diff
@@ -0,0 +1,425 @@
...
+        u_n = u_n.value(self.equation_system)
+
+        # Checking convergence specifically on the fracture
+        u_a = self.exact_sol.exact_sol_fracture(self.mdg)[1]
```
[the remark this answers is not in the document]
Comment (IvarStefansson) [reply tests/functional/setups/manu_sneddon_2d.py:369]: Please don't hard-code anything. I think we should allow the user to control this. Can be achieved by having a separate helper method and a parameter controlling the radius being removed around the tips.

File: tests/functional/test_sneddon_2d.py (hunk 19 above)
Comment (IvarStefansson) [reply tests/functional/test_sneddon_2d.py:21]: I think I don't understand exactly what your argument is. Still, I assume you're right, so if you feel confident, please move ahead. Otherwise, feel free to consult me.

File: tests/functional/test_sneddon_2d.py [hunk 26]
```diff
@@ -0,0 +1,107 @@
...
+        theta_rad=theta_rad, a=a, height=height, length=length
+    )
+
+    # Convergence analysis setup
+    conv_analysis = ConvergenceAnalysis(
+        model_class=ManuSneddonSetup2d,
+        model_params=copy.deepcopy(params),
+        levels=2,
+        spatial_refinement_rate=2,
+        temporal_refinement_rate=1,
```
[the remark this answers is not in the document]
Comment (IvarStefansson) [reply tests/functional/test_sneddon_2d.py:94]: True. The default is one, so we might as well remove it.

File: tests/functional/test_sneddon_2d.py (hunk 18 above)
Comment (IvarStefansson) [reply tests/functional/test_sneddon_2d.py:104]: Probably compatibility with other functional tests. Please have a look and assess whether it makes sense.

**Review by IvarStefansson — COMMENTED** [review]
Thanks for your efforts, @Yuriyzabegaev and @vlipovac. I have made some comments. Please implement everything you feel confident about, possibly consulting eachother and me.

File: tests/functional/test_sneddon_2d.py [hunk 27]
```diff
@@ -0,0 +1,107 @@
...
+        "domain_size": height,  # Length of square domain
+        "p0": 1e-4,  # Internal pressure of fracture
+        "poi": poi,  # Possion ratio (Not standard in solid constants)
+        "meshing_arguments": {"cell_size": 0.03},
+        "grid_type": "simplex",
+        "theta": theta_rad,
+    }
+
+    # Convert angle to radians and compute fracture points
+    params["frac_pts"] = compute_frac_pts(
```
[the remark this answers is not in the document]
Comment (Yuriyzabegaev) [reply tests/functional/test_sneddon_2d.py:84]: Moved it, please check `set_fractures`.

File: tests/functional/setups/manu_sneddon_2d.py (hunk 8 above)
Comment (Yuriyzabegaev) [reply tests/functional/setups/manu_sneddon_2d.py:314]: Changed these lines to `internal_to_dirichlet`

File: tests/functional/setups/manu_sneddon_2d.py [hunk 28]
```diff
@@ -0,0 +1,425 @@
...
+        bem_centers = get_bem_centers(self.a, h, n, self.theta, center)
+        eta = compute_eta(bem_centers, center)
+
+        u_a = -analytical_displacements(
+            self.a, eta, self.p0, self.shear_modulus, self.poi
+        )
+        u_bc = assign_bem(sd, h / 2, box_faces, self.theta, bem_centers, u_a, self.poi)
+        return u_bc
+
+    def exact_sol_fracture(
```
[the remark this answers is not in the document]
Comment (Yuriyzabegaev) [reply tests/functional/setups/manu_sneddon_2d.py:257]: Fixed

File: tests/functional/setups/manu_sneddon_2d.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (Yuriyzabegaev) [reply tests/functional/setups/manu_sneddon_2d.py:397]: Fixed

File: tests/functional/setups/manu_sneddon_2d.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (vlipovac) [reply tests/functional/setups/manu_sneddon_2d.py]: Moved to exact-solution-class and renamed according to Yuri's suggestion.

File: tests/functional/setups/manu_sneddon_2d.py [hunk 29]
```diff
@@ -0,0 +1,458 @@
...
+        bem_centers = self.get_bem_centers(h, center)
+        eta = point_pointset(bem_centers, center)
+
+        u_a = -self.analytical_displacements(eta)
+        u_bc = self.displacement_bc_values_from_bem(
+            sd, h / 2, box_faces, bem_centers, u_a
+        )
+        return u_bc
+
+    def exact_sol_fracture(self, mdg: pp.MixedDimensionalGrid) -> np.ndarray:
```
Comment (IvarStefansson) [inline tests/functional/setups/manu_sneddon_2d.py:61]: Can't we remove argument, since the class has mdg as an attribute?

File: tests/functional/setups/manu_sneddon_2d.py [hunk 30]
```diff
@@ -0,0 +1,458 @@
...
+                grid, bound_faces, new_bound_face_centers, h, u_a[i]
+            )
+
+            bc_val += u_bound
+
+        return bc_val
+
+
+class ManuSneddonGeometry2d(SquareDomainOrthogonalFractures):
+    """Square domain but with single line fracture."""
```
Comment (IvarStefansson) [inline tests/functional/setups/manu_sneddon_2d.py:251]:
```suggestion
    """Square domain  with a single, arbitrarily oriented fracture at the center."""
```

**Review by IvarStefansson — APPROVED** [review]
Thanks for the effort! The old conversations may all be resolved as far as I am concerned

File: tests/functional/setups/manu_sneddon_2d.py (hunk 29 above)
Comment (vlipovac) [reply tests/functional/setups/manu_sneddon_2d.py:61]: The exact solution is actually an attribute of the model class, not a mixin.
We could make it save a reference to the mdg upon instantiation as an alternative, but so far it saves only numerical parameters there.

Comment (IvarStefansson) [reply tests/functional/setups/manu_sneddon_2d.py:61]: Right. Keep as is, then!
